share/models/fields.py

```diff
@@ -97,66 +105,137 @@ def validate(self, value, model_instance):
 
-        version.contribute_to_class(cls, name + '_version', **kwargs)
+        version = self.PRETENDING_TO_BE(version_model, **self._get_kwargs(cls, version=True))
 
-        actual._share_version_field = version
+        suffix = '_version'
+        if self.many_to_many:
+            suffix += 's'
+            name = name.rstrip('s')
+
+        version.contribute_to_class(cls, name + suffix, **kwargs)
 
+        actual._share_version_field = version
         self.add_equivalent_fields(actual, version)
 
+    def _get_kwargs(self, cls, version=False):
+        kwargs = {**self._kwargs}
+        through_fields = kwargs.get('through_fields', None)
 
-class ShareForeignKey(ShareRelatedField, models.ForeignKey):
+        if version:
+            kwargs['db_index'] = False
+            kwargs['editable'] = False
+            kwargs['related_name'] = '+'
 
-    def __init__(self, model, **kwargs):
-        self._kwargs = kwargs
-        super().__init__(model, **kwargs)
+        if through_fields:
+            kwargs['through_fields'] = (
+                '{}_version'.format(through_fields[0]) if 'version' in cls._meta.model_name else through_fields[0],
+                '{}_version'.format(through_fields[1]) if version else through_fields[1]
+            )
+        return kwargs
 
-    def contribute_to_class(self, cls, name, **kwargs):
-        actual = self.__class__.mro()[2](self.remote_field.model, **self._kwargs)
-        actual.contribute_to_class(cls, name, **kwargs)
+    def __eq__(self, other):
+        if other in self.__equivalent_fields:
+            return True
+        return super().__eq__(other)
 
-        if isinstance(self.remote_field.model, str):
-            version = self.__class__.mro()[2](self.remote_field.model + 'Version', editable=False, **{**self._kwargs, 'related_name': '+', 'db_index': False})
-        else:
-            version = self.__class__.mro()[2](self.remote_field.model.VersionModel, editable=False, **{**self._kwargs, 'related_name': '+', 'db_index': False})
+    def __hash__(self):
+        return super().__hash__()
 
-        version.contribute_to_class(cls, name + '_version', **kwargs)
 
-        actual._share_version_field = version
+class ShareOneToOneField(ShareRelatedField, models.OneToOneField):
+    PRETENDING_TO_BE = models.OneToOneField
 
-        self.add_equivalent_fields(actual, version)
+    def __init__(self, *args, **kwargs):
+        # Default to delete cascade
+        kwargs.setdefault('on_delete', DATABASE_CASCADE)
+        super().__init__(*args, **kwargs)
+
+
+class ShareForeignKey(ShareRelatedField, models.ForeignKey):
+    PRETENDING_TO_BE = models.ForeignKey
 
+    def __init__(self, *args, **kwargs):
+        # Default to delete cascade
+        kwargs.setdefault('on_delete', DATABASE_CASCADE)
+        super().__init__(*args, **kwargs)
 
-class TypedManyToManyField(ShareRelatedField, models.ManyToManyField):
+
+def create_many_to_many_intermediary_model(field, klass):
```

Comment:
Does the generated through model differ from Django's in any way other than `on_delete=DATABASE_CASCADE`?